app/code/Magento/Catalog/Ui/DataProvider/Product/Form/Modifier/General.php

```diff
@@ -228,6 +234,7 @@ protected function customizeWeightField(array $meta)
                     'componentType' => Form\Field::NAME,
                     'dataScope' => 'product_has_weight',
                     'label' => '',
+                    'sortOrder' => $sortOrder + 1,
```

Comment:
Does this fix guarantee correct order if we have other elements with `sortOrder` and `sortOrder + 1` values?

Reply:
Usually when we drag and "sort" product attributes on the attribute set form, they are assigned sort orders with a difference of 10 (60, 70, 80,...). So unless some extension explicitly wrote a form modifier and assigned the sort order to be the same as weight or weight + 1, I think it shouldn't be a problem

Reply:
@jayankaghosh fix should guarantee order even when numbers are not consecutive.